[21277] Allow automatic creation of manageable repositories through a setting

oliverguenther opened the pull request:
This commit allows manageable repositories to be automatically created
by default whenever a new project is created and the repository module
is activated.

Relevant work package: https://community.openproject.org/work_packages/21277

oliverguenther committed:
Allow automatic creation of manageable repositories through a setting

This commit allows manageable repositories to be automatically created
by default whenever a new project is created and the repository module
is activated.

Relevant work package: https://community.openproject.org/work_packages/21277

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-automatic-repository-creation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-automatic-repository-creation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented on config/settings.yml:
```diff
@@ -137,6 +137,10 @@ sys_api_key:
 repository_authentication_caching_enabled:
   default: 1
+repositories_automatic_managed_repos:
+  default: 0
```
Would you mind adding a `type: bool`, even though this has [no effect](https://github.com/opf/openproject/pull/3270) today?

oliverguenther replied:
Ah, sure. I wasn't sure whether bool was working for the current functionality, as no other setting seemed to use it.

NobodysNightmare replied:
That is correct. It will not have any effect, besides making it clear that you intend to use this as a boolean setting :)

NobodysNightmare commented on app/assets/javascripts/settings.js.erb:
```diff
@@ -27,8 +27,18 @@ See doc/COPYRIGHT.rdoc for more details.
+
+    $('#settings_repositories_automatic_managed_repos').change(function() {
+      var checked = this.checked;
+      $('#settings_repositories_automatic_managed_vendor').prop('disabled', function() {
```
Apparently I can't see a visual difference between an enabled and a disabled select box.
This is quite annoying since I clicked three times on the select box until I realized that I need to check that checkbox before.

How about providing a `disabled` option in the select box and not offering the checkbox at all?

E.g.

> Choose an SCM type for automatic creation
> - do not create repos automatically
> - Git
> - SVN

oliverguenther replied:
Hmm, I think this is rather a question of changing the disabled style of the select, and the rest is matter of taste - I prefer having a separated switch rather.

NobodysNightmare commented:
Observation:

When setting the SCM type to be created to something invalid (or something that is not configured for managed anymore), it seems that project creation succeeds and the repository is silently not created.

Is this the desired behaviour? (I could imagine both possible answers to this question)

oliverguenther commented:
Desired? No way. But is there a sophisticated callback upon settings update? I really wouldn't want to address this in the settings view.

NobodysNightmare commented:
This one does probably not belong here. But I want to write it down before leaving the office today:

![notification](https://cloud.githubusercontent.com/assets/453584/9309455/02b2de0e-450a-11e5-9e14-f6d2cad630e9.png)

The notification overlaps the opened menu.

@oliverguenther I guess you are not even the one responsible for fixing this, are you? ^^

@floriank Any chance that you know a fix?

oliverguenther commented:
I'm going to assume this has something to do with `z-index: 1000;` :-) Continue discussion here: https://community.openproject.org/work_packages/21289

oliverguenther committed:
Refactor automatic repository setting into single select

This commit moves the automatic repository setting into a single
select and moves it below the enabled vendors.

Changing the enabled vendors will now also dynamically add or remove vendors
from the repository select field.
The select field is disabled when no option remains.

oliverguenther commented:
> When setting the SCM type to be created to something invalid (or something that is not configured for managed anymore), it seems that project creation succeeds and the repository is silently not created.

Concerning this problem, I've let you change my mind concerning the two options. I've refactored it into a single select residing below the *Enabled SCMs multi-checkbox. The select field will dynamically gain / lose options changed in that multi-checkbox and disable itself when no option remains.

![bildschirmfoto 2015-08-18 um 08 59 00](https://cloud.githubusercontent.com/assets/459462/9324113/83ebb054-4587-11e5-925e-4a946648c8df.png)

Azure7111 commented:
You got "automatically" twice in a row in your screenshot ;)

oliverguenther commented:
Obviously, it's highly automatically and deserves a double mention! :grinning: It's correct in the PR field.

oliverguenther committed:
Enable SCM config for spec

NobodysNightmare commented on app/assets/javascripts/settings.js.erb:
```diff
@@ -27,8 +27,42 @@ See doc/COPYRIGHT.rdoc for more details.
+        select = $('#settings_repositories_automatic_managed_vendor'),
+        option = select.find('option[value="' + value + '"]'),
+        makeOption = function() {
+          return ['<option value="', value, '">', value, '</option>'].join('');
```
Is this the best way to do string concatenation in JS?
Why not use good old addition of strings?

oliverguenther replied:
I somehow remember it to be faster, however that information seems to be wrong (or outdated). I've changed it along with the r4 rebase.

oliverguenther committed:
Use String concat instead of Array join

NobodysNightmare commented:
LGTM :+1: 

Waiting for travis to agree on that.
